internal/expr/parser.go

```diff
@@ -135,24 +144,73 @@ func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
 			p.add(in)
 			continue
 		}
+	}
 
-		if bypass, ok, err := p.parseStringLiteral(); err != nil {
-			return nil, err
+	// Add any remaining unparsed string input to the parser.
+	p.partStart = p.pos
+	p.add(nil)
+	return &ParsedExpr{p.parts}, nil
+}
+
+// advance increments p.pos until it reaches content that might preceed a token
+// we want to parse.
+func (p *Parser) advance() error {
+
+	// The byte following these bytes might occur before the start of an expression.
+	delimiterBytes := map[byte]bool{
+		' ':  true,
+		'\t': true,
+		'\n': true,
+		'\r': true,
+		')':  true,
+		';':  true,
+		'=':  true,
+		',':  true,
+	}
+
+	for p.pos < len(p.input) {
+		if ok, err := p.skipStringLiteral(); err != nil {
+			return err
 		} else if ok {
-			p.add(bypass)
 			continue
 		}
 
-		if p.pos == len(p.input) {
+		if delimiterBytes[p.input[p.pos]] {
+			p.pos++
 			break
 		}
-
-		// If nothing above can be parsed we advance the parser.
 		p.pos++
 	}
-	// Add any remaining unparsed string input to the parser.
-	p.add(nil)
-	return &ParsedExpr{p.parts}, nil
+
+	p.skipBlanks()
+
+	return nil
+
+}
+
+// skipStringLiteral jumps over single and double quoted sections of input.
+// Doubled up quotes are escaped.
+func (p *Parser) skipStringLiteral() (bool, error) {
+	mark := p.pos
+	if p.skipByte('"') || p.skipByte('\'') {
```

Comment:
I can't see how it would corrupt the state unless I'm missing something. On false both of the `p.skipByte` have returned false and thus not changed the state, we then return immediately.

We do return false and an error on line 201 which would result in a corrupted state but since it is an error state I assume this is ok. There are other places in the parser (`parseFullName` or `parseList`) where we return with an error and a corrupted state.

Reply:
It's not okay, even if there is an error result, because it's up to the call site to decide if that's a fatal error or not. If any of these functions don't save/restore, though, we take away the ability of the call site to decide for itself if errors are final or if there's a an alternative to be attempted for the given scenario.

It's also much easier to keep a strong invariant in mind than doing pinpointed decisions everywhere. The invariant here is: if it fails the state is untouched.

Reply:
Ok, I've changed it. I'll create a new PR to preserve it in other places to.